### ISSUE TITLE
slow routing rule sync for borrowed IP when a node is failed

### ISSUE THREAD

**Issue by WangWooLee** [projectcalico/calico#4179]
## Issue

We want pod failover which pod has static IP, and without any delays. To create pod using static IP, we created IP pool and wrote pod's spec like this :

```
apiVersion: projectcalico.org/v3
kind: IPPool
metadata:
  name: ippool-test
spec:
  cidr: 172.22.22.100/31
  blockSize: 31
  ipipMode: Never
  natOutgoing: true
  disabled: false
  nodeSelector: all()
```

```
apiVersion: v1
kind: Pod
metadata:
  name: hello-world
  labels:
    app: hello-world-example1
  annotations:
     "cni.projectcalico.org/ipAddrs": "[\"172.22.22.100\"]"
```

So our pod has static IP - 172.22.22.100 , and IP pool CIDR is /31.
172.22.22.100 IP block is assigned to Worker-1. so routing table looks like : 

```
- Master node
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
172.22.22.100   10.10.1.11(Worker-1)      255.255.255.254 UG    0      0        0 eth1

- Worker-1
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
172.22.22.100   0.0.0.0         255.255.255.255 UH    0      0        0 cali1f818560fa0
172.22.22.100   0.0.0.0         255.255.255.254 U     0      0        0 *(blackhole)

- Worker-2
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
172.22.22.100   10.10.1.11(Worker-1)      255.255.255.254 UG    0      0        0 eth1
```
Worker-1 node has "172.22.22.100/31 -> blackhole" and "172.22.22.100/32 -> calixxxxxx" routing rule, and another nodes have "172.22.22.100/31 -> Worker-1" rule.

#### First failover

in this situation, we shut downed Worker-1 node, and release IP(172.22.22.100) for Pod failover to another node. we can create pod with same IP immediately, without any error.

```
$kubectl get pods -o wide
NAME           READY   STATUS        RESTARTS   AGE     IP              NODE       NOMINATED NODE   READINESS GATES
hello-world    1/1     Terminating   0          6m57s   172.22.22.100   worker-1   <none>           <none>
hello-world2   1/1     Running       0          3s      172.22.22.100   worker-2   <none>           <none>
```

there is pod with static IP on terminating, but we can create pod with same IP after "calicoctl ipam release --ip=172.22.22.100)". new pod is on Worker-2 node. "hello-world" pod will be cleanly terminated when Worker-1 node recovered. after Worker-1 node is recovered, the routing table is  :

```
- Master node
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
172.22.22.100   10.10.1.12(Worker-2)      255.255.255.255 UGH   0      0        0 eth1
172.22.22.100   10.10.1.11(Worker-1)      255.255.255.254 UG    0      0        0 eth1

- Worker-1
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
172.22.22.100   10.10.1.12(Worker-2)      255.255.255.255 UGH   0      0        0 eth1
172.22.22.100   0.0.0.0         255.255.255.254 U     0      0        0 *

- Worker-2
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
172.22.22.100   0.0.0.0         255.255.255.255 UH    0      0        0 cali5db851776c1
172.22.22.100   10.10.1.11(Worker-1)      255.255.255.254 UG    0      0        0 eth1
```

as you see, 172.22.22.100/32 is borrowed to Worker-2 from Worker-1. so on Worker-1's routing table, there are /31 -> *(black hole) , /32 -> Worker-2  rules. and on Worker-2, /31 -> Worker-1, /32 -> calixxxxx.
on Master node, /31 -> Worker-1, which IP block is assigned at first time, and /32 -> Worker-2, which is borrowing node. we could check it's borrowed on calicoctl.

```
$ calicoctl ipam show --show-borrowed
+---------------+----------------+------------------+-------------+------+----------------------+
|      IP       | BORROWING-NODE |      BLOCK       | BLOCK OWNER | TYPE |     ALLOCATED-TO     |
+---------------+----------------+------------------+-------------+------+----------------------+
| 172.22.22.100 | worker-2       | 172.22.22.100/31 | worker-1    | pod  | default/hello-world2 |
+---------------+----------------+------------------+-------------+------+----------------------+
```

#### second failover

We shut downed Worker-2  to see second failover. we released IP, and create pod with same IP. but it had error unlike the first failover : 
```
$kubectl describe pod hello-world
Events:
  Type     Reason                  Age               From               Message
  Normal   Scheduled               15s               default-scheduler  Successfully assigned default/hello-world to worker-1

  Warning  FailedCreatePodSandBox  14s               kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "e284ffffa06f3ef69445eca3f4b0a2b8308e26acdf96b7d4a30726f22407d0e2" network for pod "hello-world": networkPlugin cni failed to set up pod "hello-world_default" network: error adding host side routes for interface: cali1f818560fa0, error: route (Ifindex: 8, Dst: 172.22.22.100/32, Scope: 253) already exists for an interface other than 'cali1f818560fa0'
```

This error appeared when /32 rule(172.22.22.100/32 -> Worker-2) is already on scheduled node. after a few seconds(1m30s I guess), /32 routing rule on worker-1 is removed with this log on calico-node : 

```
bird: Mesh_10_10_1_11: State changed to stop
bird: Mesh_10_10_1_11: State changed to down
bird: Mesh_10_10_1_11: Starting
bird: Mesh_10_10_1_11: State changed to start
```

we didn't want any delays on second failover, so we tried several ways to remove /32 routing rule about already dead pod,  like deleting pod with --force option, deleting calico workloadendpoint, ... but only bird restart (by birdcl > restart all) works for deleting /32 rule immediately.
on first failover, it can failover without any delays, because there is no /32 rule on other nodes. but second failover, there is /32 rule inserted by BIRD on every nodes. so must wait BIRD remove /32 rule few seconds later after node fail. I guess BIRD or Felix on failed node should do something but it couldn't when node is totally shut downed. actually it works without any delays on second failover when BIRD and Felix are alive on fail node.
So we want pod failover using static IP without any delays, and multiple times. any better idea except BIRD restart on calico-node? at least I want to find some easy way to restart it.

Thank you for reading it.

## Your Environment

* Calico version : v3.16.5
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes v1.19.4
* Operating System and version: Ubuntu 18.04
* Link to your project (optional):

**Comment by caseydavenport** [comment]
Yeah, BIRD is programming that route and so BIRD will need to remove it. It's learning the route from its BGP session with the other node, and if the other node goes down there is a grace period for BGP graceful restart that will keep the route around waiting for the remote node to come back online. 

Forcing the restart is likely bypassing the graceful restart handling, and thus forcing BIRD to update its routing information despite its reluctance to ignore the downed node.

I'm not sure what the right answer here is  - if you don't need BGP, one option might be to use the VXLAN backend which doesn't rely on BGP route distribution, and might respond faster to IP addresses moving in this way. 

You could also try gracefully terminating the remote node, or deleting the node (e.g., `kubectl delete node`) so that the other nodes know it no longer exists, which should trigger them to remove any routes they learned from it quicker.

**Comment by caseydavenport** [comment]
Looking at this again, have you tried using a Kubernetes service to implement this? That's sort of what they are intended to do - handle moving pods. 

Generally it's a bad / unexpected practice to use IP mobility as a means of failover in Kubernetes - instead, use a Service and it will ensure that your traffic reaches a pod independent of what IP is ultimately assigned to the pod interface.
